compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs

```diff
@@ -226,10 +226,11 @@ impl<'tcx> ConstToPat<'tcx> {
         // using `PartialEq::eq` in this scenario in the past.)
         let partial_eq_trait_id =
             self.tcx().require_lang_item(hir::LangItem::PartialEq, Some(self.span));
+        let def_id = self.tcx().hir().local_def_id(self.id);
```

Comment:
This is causing to me some problems, and I do not think I should use the body id here? the const should have a LocalDefId most of the time right?

```diff
+++ b/compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs
@@ -1,3 +1,4 @@
+use hir::def_id::CRATE_DEF_ID;
 use rustc_errors::DelayDm;
 use rustc_hir as hir;
 use rustc_index::vec::Idx;
@@ -226,7 +227,7 @@ fn type_may_have_partial_eq_impl(&self, ty: Ty<'tcx>) -> bool {
         // using `PartialEq::eq` in this scenario in the past.)
         let partial_eq_trait_id =
             self.tcx().require_lang_item(hir::LangItem::PartialEq, Some(self.span));
-        let def_id = self.tcx().hir().local_def_id(self.id);
+        let def_id = self.tcx().hir().opt_local_def_id(self.id).unwrap_or(CRATE_DEF_ID);
         let obligation: PredicateObligation<'_> = predicate_for_trait_def(
             self.tcx(),
             self.param_env,

```